### DISCUSSION TITLE
Помогите с маркировкой пакетов iptables

### DISCUSSION THREAD

**Discussion by Anonym-tsk** [bol-van/zapret#782] [Q&A]
Дано - маршрутизатор с возможностью создавать политики и назначать их клиентам.
При добавлении клиента в политику, его трафик маркируется примерно таким правилом
```
-A _NDM_HOTSPOT_PRERT -m mac --mac-source 02:11:32:2B:56:01 -j MARK --set-xmark 0xffffaac/0xffffffff
```
Однако NFQWS тоже маркирует трафик меткой `0x40000000`

**Задача 1**
Хочу сделать политику и чтобы NFQWS работал только для клиентов этой политики.
Что имеем:
- Роутер маркирует трафик меткой `0xffffaac/0xffffffff`
- Для NFQWS пишем примерно такое правило
```
-A POSTROUTING -o eth3 -t mangle -p tcp -m multiport --dports 443,80 -m connbytes --connbytes-dir=original --connbytes-mode=packets --connbytes 1:8 -m mark --mark 0xffffaac -j NFQUEUE --queue-num 200 --queue-bypass
```
- После прохождения пакета через NFQWS, он перемаркировывается меткой `0x40000000`, но вроде все работает как надо.

**Задача 2**
Хочу сделать политику и чтобы NFQWS _не работал_ для клиентов этой политики.

- Роутер маркирует трафик меткой `0xffffaac/0xffffffff`
- Для NFQWS пишем примерно такое правило
```
-A POSTROUTING -o eth3 -t mangle -p tcp -m multiport --dports 443,80 -m connbytes --connbytes-dir=original --connbytes-mode=packets --connbytes 1:8 -m mark ! --mark 0xffffaac -j NFQUEUE --queue-num 200 --queue-bypass
```
- После прохождения пакета через NFQWS, он перемаркировывается меткой `0x40000000`, и сможет повторно попасть опять на обработку.

Помогите скрестить одно с другим, пожалуйста.

**Comment by bol-van** [comment]
Начнем с того, что nfqws использует только 1 бит, и он нужен для предотвращения зацикливания и избегания дедлоков.
Он работает с пакетами в очереди. Берет пришедший марк и ORит его с битом desync mark.
	uint32_t desync_fwmark = fwmark | params.desync_fwmark;
Другие биты он не трогает.
Этот марк используется для сгенерированных пакетов. Цепочка OUTPUT, с самого роутера, не откуда-то пришедших.
nfqws может дропать в очереди приходящие откуда-то пакеты. Они перестают быть, исчезают из системы вместе с их маркировкой. Но вместо них генерятся свои пакеты в цепочке OUTPUT с марком OR desync mark.
Вот из этой логики и надо исходить.
Если прошивка где-то занимает весь марк 0xffffffff, то это плохо

**Comment by ebkjiud** [comment]
Возник вопрос о запуске несколько процессов NFQWS и разруливание их через iptables
Если я правильно понимаю
-A INPUT -i enp2s0 -p udp -m multiport --sports 443 -m connbytes --connbytes 1:55 --connbytes-mode packets --connbytes-dir reply -m set ! --match-set nozapret src -j NFQUEUE --queue-num 300 --queue-bypass

Это правило сверху подразумевает входящий пакет по конкретному conntract, который в свою очередь помечен меткой xmark через правило ниже:
-A POSTROUTING -m mark --mark 0x40000000/0x40000000 -j CONNMARK --set-xmark 0x40000000/0x40000000

Вопрос:
Нужно ли помечать входящий трафик чтобы он попал в свою очередь же --queue-num?

Как я понимаю мне требуется в prerouting добавить правило, что все пакеты которые проходят conntract с xmark метилось без пересечения меткой пакета, и уже на этапе INPUT и FORWARD ответа по метке пакета попадал в нужную очередь, главное чтобы эта метка не совпадала, с той меткой которую ставит NFQWS?
а разные процессы NFQWS имели разную метку пакетов и метку conntract (mark и xmark)

И наверно проще делать цепочки отдельные для разных процессов?


nfqws2 --fwmark=0x20000000 --qnum=200
nfqws2 --fwmark=0x30000000 --qnum=300

-A POSTROUTING -m mark --mark 0x20000000/0x20000000 -j CONNMARK --set-xmark 0x20000000/0x20000000
-A POSTROUTING -m mark --mark 0x30000000/0x20000000 -j CONNMARK --set-xmark 0x30000000/0x30000000
-A POSTROUTING -o enp2s0 -p udp -m mark --mark 0x20000000/0x20000000 -m multiport --dports 443 -m connbytes --connbytes 1:55 --connbytes-mode packets --connbytes-dir original -m set ! --match-set nozapret dst -j ACCEPT
-A POSTROUTING -o enp2s0 -p udp -m mark --mark 0x30000000/0x30000000 -m multiport --dports 443 -m connbytes --connbytes 1:55 --connbytes-mode packets --connbytes-dir original -m set ! --match-set nozapret dst -j ACCEPT

-A POSTROUTING с нужными условиями в нужную очередь

-A PREROUTING -j CONNMARK --restore-mark

-A INPUT -m mark --mark 0x20000000/0x20000000 -//- -j NFQUEUE --queue-num 200 --queue-bypass
-A INPUT -m mark --mark 0x30000000/0x30000000 -//- -j NFQUEUE --queue-num 300 --queue-bypass

**Reply by bol-van** [reply]
nfqws помечает --fwmark только генерируемые им пакеты. Это нужно для предовтращения loop. Можно использовать 1 бит на все инстансы.
Каждый инстанс должен получат все пакеты одного соединения, разброс их по разным инстансам приведет к неправильной работе. Эту задачу и нужно решить через connmark, но вне связи с fwmark nfqws.

**Reply by ebkjiud** [reply]
ну т.е. идея с "-A PREROUTING -j CONNMARK --restore-mark" верна, только наверно мне требуется указать условие -m connbytes --connbytes 1:55 --connbytes-mode packets --connbytes-dir reply -m set ! --match-set nozapret src
чтобы только ответные пакеты востанавливали метку и по метки их заруливать в требуемую queue-num которой полбьзуется процесс

**Reply by bol-van** [reply]
CONNMARK выставляется один раз, можно это повесить на --syn или на connbytes `:1` - так и на udp должно работать
по каким условиям - вам решать. если это разные инстансы с разными стратегиями - одни условия. например, source ip или lan interface или еще что-то. если балансировка - надо плясать от чего-то вроде младших битов src port, но как это делать на iptables не изучал. на nft проще.
на остальные пакеты - не SYN - на все правила nfqueue добавить проверку -m connmark --mark x/y. если такой - на одну очередь, если другой - на другую. connbytes тоже проверять, чтобы не тащить все с ядра.
как-то так в теории это видится, я не проверял

**Reply by ebkjiud** [reply]
разобрался спасибо, чуть запутался с метками и их работой

**Reply by bol-van** [reply]
Вот еще вспомнил. Есть же ctstate NEW. Можно по нему ставить connmark
